capsules/src/kv_driver.rs

```diff
@@ -0,0 +1,467 @@
+//! KV Driver
+//!
+
+use crate::driver;
+/// Syscall driver number.
+pub const DRIVER_NUM: usize = driver::NUM::KVSystem as usize;
+
+use crate::kv_store::KVStore;
+use core::cell::Cell;
+use kernel::grant::Grant;
+use kernel::grant::{AllowRoCount, AllowRwCount, UpcallCount};
+use kernel::hil::kv_system;
+use kernel::processbuffer::{ReadableProcessBuffer, WriteableProcessBuffer};
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::utilities::cells::{OptionalCell, TakeCell};
+use kernel::{ErrorCode, ProcessId};
+
+/// Ids for read-only allow buffers
+mod ro_allow {
+    // unhashed key
+    pub const UNHASHED_KEY: usize = 0;
+    // input value
+    pub const VALUE: usize = 1;
+    /// The number of allow buffers the kernel stores for this grant
+    pub const COUNT: usize = 2;
+}
+
+/// Ids for read-write allow buffers
+mod rw_allow {
+    pub const VALUE: usize = 0;
+    /// The number of allow buffers the kernel stores for this grant
+    pub const COUNT: usize = 1;
+}
+
+pub struct KVSystemDriver<
+    'a,
+    K: kv_system::KVSystem<'a> + kv_system::KVSystem<'a, K = T>,
+    T: 'static + kv_system::KeyType,
+> {
+    kv: &'a KVStore<'a, K, T>,
+
+    active: Cell<bool>,
+
+    apps: Grant<
+        App,
+        UpcallCount<1>,
```

Comment:
I am a fan of the ro_allow::COUNT approach, could we do that for the upcall as well here for consistency?